rest/blip_sync_messages.go

```diff
@@ -0,0 +1,229 @@
+package rest
+
+import (
+	"fmt"
+	"math"
+	"strings"
+	"bytes"
+
+	"github.com/couchbase/go-blip"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/couchbase/sync_gateway/channels"
+	"github.com/couchbase/sync_gateway/db"
+)
+
+// Function signature for something that generates a sequence id
+type SequenceIDGenerator func() db.SequenceID
+
+// Helper for handling BLIP subChanges requests.  Supports Stringer() interface to log aspects of the request.
+type subChanges struct {
+	rq                    *blip.Message       // The underlying BLIP message
+	logger                base.SGLogger       // A logger object which might encompass more state (eg, blipContext id)
+	sinceZeroValueCreator SequenceIDGenerator // A sequence generator for creating zero'd since values
+}
+
+// Create a new subChanges helper
+func newSubChanges(rq *blip.Message, logger base.SGLogger, sinceZeroValueCreator SequenceIDGenerator) *subChanges {
+	return &subChanges{
+		rq:                    rq,
+		logger:                logger,
+		sinceZeroValueCreator: sinceZeroValueCreator,
+	}
+}
+
+func (s *subChanges) since() db.SequenceID {
+
+	// Depending on the db sequence type, use correct zero sequence for since value
+	sinceSequenceId := s.sinceZeroValueCreator()
+
+	if sinceStr, found := s.rq.Properties["since"]; found {
+		var err error
+		if sinceSequenceId, err = db.ParseSequenceIDFromJSON([]byte(sinceStr)); err != nil {
+			s.logger.LogTo("Sync", "%s: Invalid sequence ID in 'since': %s", s.rq, sinceStr)
+			sinceSequenceId = db.SequenceID{}
+		}
+	}
+
+	return sinceSequenceId
+
+}
+
+func (s *subChanges) batchSize() int {
+	return int(getRestrictedIntFromString(s.rq.Properties["batch"], BlipDefaultBatchSize, 10, math.MaxUint64, true))
+}
+
+func (s *subChanges) continuous() bool {
+	continuous := false
+	if val, found := s.rq.Properties["continuous"]; found && val != "false" {
+		continuous = true
+	}
+	return continuous
+}
+
+func (s *subChanges) activeOnly() bool {
+	return (s.rq.Properties["active_only"] == "true")
+}
+
+func (s *subChanges) filter() string {
+	return s.rq.Properties["filter"]
+}
+
+func (s *subChanges) channels() (channels string, found bool) {
+	channels, found = s.rq.Properties["channels"]
+	return channels, found
+}
+
+func (s *subChanges) channelsExpandedSet() (resultChannels base.Set, err error) {
+	channelsParam, found := s.rq.Properties["channels"]
+	if !found {
+		return nil, fmt.Errorf("Missing 'channels' filter parameter")
+	}
+	channelsArray := strings.Split(channelsParam, ",")
+	return channels.SetFromArray(channelsArray, channels.ExpandStar)
+}
+
+// Satisfy fmt.Stringer interface for dumping attributes of this subChanges request to logs
+func (s *subChanges) String() string {
+
+	buffer := bytes.NewBufferString("")
+
+	buffer.WriteString(fmt.Sprintf("Since:%v ", s.since()))
```

Comment:
s.since() looks like it recalculates since (and potentially logs errors on invalid since).  That doesn't feel like the right thing to be doing in String().

Reply:
More generally, this makes me wonder if we should just be building the logging string itself in handleSubChanges, instead of repeating all the request parameter parsing at log time.  

In particular, we're not trying to log a the state of mutable properties - we're just trying to log what the incoming request looked like.

Reply:
>  this makes me wonder if we should just be building the logging string itself in handleSubChanges

Currently for subChanges this is ~35 lines of code:

```
// Satisfy fmt.Stringer interface for dumping attributes of this subChanges request to logs
func (s *subChanges) String() string {

	buffer := bytes.NewBufferString("")

	since, err := s.since()
	if err != nil {
		base.Warn("Error discovering since value from subchanges.  Err: %v", err)
	} else {
		buffer.WriteString(fmt.Sprintf("Since:%v ", since))
	}

	continuous := s.continuous()
	if continuous {
		buffer.WriteString(fmt.Sprintf("Continuous:%v ", continuous))
	}

	activeOnly := s.activeOnly()
	if activeOnly {
		buffer.WriteString(fmt.Sprintf("ActiveOnly:%v ", activeOnly))
	}

	filter := s.filter()
	if len(filter) > 0 {
		buffer.WriteString(fmt.Sprintf("Filter:%v ", filter))
		channels, found := s.channels()
		if found {
			buffer.WriteString(fmt.Sprintf("Channels:%v ", channels))
		}
	}

	batchSize := s.batchSize()
	if batchSize != int(BlipDefaultBatchSize) {
		buffer.WriteString(fmt.Sprintf("BatchSize:%v ", s.batchSize()))
	}

	return buffer.String()

}
```

I guess it could be condensed, but it feels like it would have to be condensed quite a bit to fit in `handleSubChanges()` while keeping `handleSubChanges()` readable.  It could be a method rather than a helper struct, but the helper struct that conforms to the `fmt.Stringer()` interface allows some additional consistency / reuse in the code